# E-commerce Sales Analysis

This notebook analyzes an online retail transaction dataset to understand revenue performance, top products, key countries, and high-value customers.

## TL;DR

- The cleaned dataset contains completed sales transactions only.
- The analysis focuses on revenue, seasonality, product categories, customer retention, and market performance.
- The outputs are saved in `reports/` so the README can show the main findings clearly.

## Context & Methods

Business goal: turn raw transaction data into useful sales insights for inventory planning, market focus, and customer targeting.

Cleaning rules:

- remove rows with missing product descriptions or customer IDs
- remove cancelled invoices
- remove zero or negative quantities
- remove zero or negative unit prices
- calculate `Revenue = Quantity * UnitPrice`

### Key Assumption

The dataset has no official product category field. Categories are inferred from documented description keywords, so category findings are directional rather than an official merchandising taxonomy.

## 1. Load Analysis Functions

In [1]:
from src.ecommerce_sales_analysis import (
    build_data_quality_report,
    build_outputs,
    clean_data,
    load_data,
    save_figures,
    save_summary,
    save_tables,
)

## 2. Load and Inspect the Dataset

In [2]:
raw_df = load_data()
raw_df.shape

(541909, 8)

In [3]:
raw_df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [4]:
raw_df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

## 3. Clean the Data

In [5]:
clean_df = clean_data(raw_df)
clean_df.shape

(397884, 12)

In [6]:
clean_df[["InvoiceNo", "Description", "Quantity", "UnitPrice", "CustomerID", "Country", "Revenue"]].head()

,InvoiceNo,Description,Quantity,UnitPrice,CustomerID,Country,Revenue
0,536365,WHITE HANGING HEART T-LIGHT HOLDER,6,2.55,17850,United Kingdom,15.30
1,536365,WHITE METAL LANTERN,6,3.39,17850,United Kingdom,20.34
2,536365,CREAM CUPID HEARTS COAT HANGER,8,2.75,17850,United Kingdom,22.00
3,536365,KNITTED UNION FLAG HOT WATER BOTTLE,6,3.39,17850,United Kingdom,20.34
4,536365,RED WOOLLY HOTTIE WHITE HEART.,6,3.39,17850,United Kingdom,20.34


## 4. Build Data Quality Report

In [7]:
quality_outputs = build_data_quality_report(raw_df, clean_df)
quality_outputs["data_quality_summary"]

,Check,AffectedRows,AffectedRate,AnalyticalRisk,Severity
0,Missing product descriptions,1454,0.002683,Cannot support reliable product-level analysis.,Medium
1,Missing customer IDs,135080,0.249267,Cannot support customer-level segmentation or ...,High
2,Cancelled invoices,9288,0.017139,"Represents returns or cancellations, not compl...",High
3,Zero or negative quantity,10624,0.019605,Would distort revenue and volume metrics for n...,High
4,Zero or negative unit price,2517,0.004645,Would distort revenue and average order value.,High
5,Exact duplicate rows,5268,0.009721,May double-count transactions if not reviewed.,Medium


In [8]:
quality_outputs["cleaning_funnel"]

,Step,RowsRemaining,RowsRemoved,RowsRemainingRate
0,Raw data,541909,0,1.000000
1,Remove missing product descriptions,540455,1454,0.997317
2,Remove zero or negative quantity,530693,9762,0.979303
3,Remove zero or negative unit price,530104,589,0.978216
4,Remove missing customer IDs,397884,132220,0.734227
5,Remove cancelled invoices,397884,0,0.734227


## 5. Calculate KPIs and Ranking Tables

In [9]:
outputs = build_outputs(clean_df)
outputs["kpis"]

,Metric,Value
0,Total revenue,8.911408e+06
1,Completed transactions,3.978840e+05
2,Unique invoices,1.853200e+04
3,Unique customers,4.338000e+03
4,Unique products,3.877000e+03
5,Countries,3.700000e+01
6,Average order value,4.808660e+02


In [10]:
outputs["monthly_revenue"]

,YearMonth,Revenue
0,2010-12,572713.890
1,2011-01,569445.040
2,2011-02,447137.350
3,2011-03,595500.760
4,2011-04,469200.361
5,2011-05,678594.560
6,2011-06,661213.690
7,2011-07,600091.011
8,2011-08,645343.900
9,2011-09,952838.382


## 6. Sales Seasonality Analysis

Seasonality analysis compares sales performance by month, week, and weekday. This helps identify when inventory, staffing, and promotions may need extra planning.

In [11]:
outputs["seasonality_summary"]

,Insight,Period,Revenue
0,Highest revenue month,2011-11,1161817.380
1,Lowest revenue month,2011-02,447137.350
2,Highest revenue week,2011-W49,408569.150
3,Lowest revenue week,2010-W51,46109.110
4,Highest revenue weekday,Thursday,1976859.070
5,Lowest revenue weekday,Sunday,792514.221


In [12]:
outputs["weekly_revenue"].head()

,YearWeek,Revenue
0,2010-W48,149386.33
1,2010-W49,213447.72
2,2010-W50,163770.73
3,2010-W51,46109.11
4,2011-W01,114865.27


In [13]:
outputs["weekday_revenue"]

,Weekday,Revenue
1,Monday,1367146.411
4,Tuesday,1700634.631
5,Wednesday,1588336.170
3,Thursday,1976859.070
0,Friday,1485917.401
2,Sunday,792514.221


In [14]:
outputs["top_products"]

,Description,Revenue
2319,"PAPER CRAFT , LITTLE BIRDIE",168469.60
2767,REGENCY CAKESTAND 3 TIER,142592.95
3698,WHITE HANGING HEART T-LIGHT HOLDER,100448.15
1762,JUMBO BAG RED RETROSPOT,85220.78
1992,MEDIUM CERAMIC TOP STORAGE JAR,81416.73
2611,POSTAGE,77803.96
2345,PARTY BUNTING,68844.33
216,ASSORTED COLOUR BIRD ORNAMENT,56580.34
2130,Manual,53779.93
2656,RABBIT NIGHT LIGHT,51346.20


In [15]:
outputs["top_countries"]

,Country,Revenue
35,United Kingdom,7308391.554
23,Netherlands,285446.340
10,EIRE,265545.900
14,Germany,228867.140
13,France,209024.050
0,Australia,138521.310
30,Spain,61577.110
32,Switzerland,56443.950
3,Belgium,41196.340
31,Sweden,38378.330


In [16]:
outputs["top_customers"]

,CustomerID,Revenue
1689,14646,280206.02
4201,18102,259657.30
3728,17450,194550.79
3008,16446,168472.50
1879,14911,143825.06
55,12415,124914.53
1333,14156,117379.63
3771,17511,91062.38
2702,16029,81024.84
0,12346,77183.60


## 7. Customer Retention Analysis

Customers are grouped as repeat customers when they have more than one unique invoice. This helps compare one-time buyers with customers who return and buy again.

In [17]:
outputs["customer_retention_summary"]

,CustomerType,Customers,Revenue,AverageRevenuePerCustomer,AverageInvoicesPerCustomer,CustomerShare,RevenueShare
1,Repeat customer,2845,8295096.173,2915.675280,5.989104,0.655832,0.93084
0,One-time customer,1493,616311.731,412.800891,1.000000,0.344168,0.06916


## 8. Product Category Analysis

Products are grouped using a transparent keyword-based taxonomy. Revenue, quantity, invoices, customers, and product breadth are compared by inferred category.

In [18]:
outputs["product_category_summary"]

,ProductCategory,Revenue,Quantity,Invoices,Customers,Products,RevenueShare,AverageRevenuePerInvoice
3,Home Decor,2376070.770,1021008,14482,3861,826,0.266632,164.070624
4,Kitchen & Dining,1933683.720,953356,14104,3882,673,0.216990,137.101795
6,Other,1747120.413,1227167,14685,3970,1205,0.196054,118.973130
0,Bags & Storage,1385906.020,857884,11221,3322,292,0.155520,123.510028
9,Stationery & Crafts,478468.190,566971,7467,2684,269,0.053692,64.077701
8,Seasonal,412501.340,303842,5577,2547,204,0.046289,73.964737
2,Gifts & Accessories,174878.560,127689,4719,2073,197,0.019624,37.058394
5,Non-product Charges,152549.821,10683,1646,644,8,0.017118,92.679114
1,Garden & Outdoor,131894.150,59040,3926,1973,135,0.014801,33.595046
10,Toys & Children,97371.240,31154,2435,1238,42,0.010927,39.988189


In [19]:
outputs["top_product_by_category"]

,ProductCategory,Description,Revenue
3701,Stationery & Crafts,"PAPER CRAFT , LITTLE BIRDIE",168469.60
1922,Kitchen & Dining,REGENCY CAKESTAND 3 TIER,142592.95
1390,Home Decor,WHITE HANGING HEART T-LIGHT HOLDER,100448.15
110,Bags & Storage,JUMBO BAG RED RETROSPOT,85220.78
2130,Non-product Charges,POSTAGE,77803.96
2355,Other,CHILLI LIGHTS,46286.51
3496,Seasonal,PAPER CHAIN KIT 50'S CHRISTMAS,42660.83
504,Gifts & Accessories,JUMBO SHOPPER VINTAGE RED PAISLEY,20174.23
3851,Toys & Children,FELTCRAFT DOLL MOLLY,11762.99
423,Garden & Outdoor,WOODEN ROUNDERS GARDEN SET,11601.14


## 9. Save Portfolio Outputs

In [20]:
outputs.update(quality_outputs)
save_tables(outputs)
save_figures(outputs)
save_summary(clean_df, outputs, quality_outputs)

print("Portfolio outputs saved in reports/.")

Portfolio outputs saved in reports/.


## Results

### Monthly Revenue Trend

![Monthly revenue trend](reports/figures/monthly-revenue.svg)

### Weekly Revenue Trend

![Weekly revenue trend](reports/figures/weekly-revenue.svg)

### Revenue by Weekday

![Revenue by weekday](reports/figures/weekday-revenue.svg)

### Top Products by Revenue

![Top products by revenue](reports/figures/top-products.svg)

### Top Countries by Revenue

![Top countries by revenue](reports/figures/top-countries.svg)

### Customer Retention

![Revenue by customer retention type](reports/figures/customer-retention.svg)

### Revenue by Product Category

![Revenue by product category](reports/figures/product-category-revenue.svg)

## Takeaways

- Revenue is concentrated in the United Kingdom, which should remain the primary market focus.
- The strongest sales period is November to early December, so inventory and promotional planning should happen before the seasonal peak.
- Thursday is the strongest weekday in the cleaned data, while Saturday does not appear in the cleaned transaction records.
- Repeat customers generate most of the revenue, so retention and loyalty actions should be a priority.
- Home Decor and Kitchen & Dining generate 48.4% of cleaned revenue and are strong candidates for merchandising and inventory tests.
- The Other group represents 19.6% of revenue, so the inferred taxonomy should be reviewed before operational use.
- High-revenue products and customers can be used for targeted stock planning, retention campaigns, and sales follow-up.